### PR TITLE
added malloc checks

### PR DESCRIPTION
**Pull request by ericwol-msft**

### CODE REVIEW

File: src/map.c [hunk 1]
```diff
@@ -222,20 +230,24 @@ static void Map_DecreaseStorageKeysValues(MAP_HANDLE_DATA* handleData)
...
+            LogError("CATASTROPHIC error, unable to undo through realloc to a smaller size, size:%zu", realloc_size);
         }
         else
         {
             handleData->keys = undoneKeys;
         }
 
-        undoneValues = (char**)realloc(handleData->values, sizeof(char*)* (handleData->count - 1));
-        if (undoneValues == NULL)
+        if (realloc_size == SIZE_MAX ||
```
Comment (RLeclair) [inline src/map.c:247]: Looks like this condition was already checked on line 237

**Review by RLeclair — APPROVED** [review]
LGTM!

File: src/map.c (hunk 1 above)
Comment (ericwol-msft) [reply src/map.c:247]: yea, there is no function exit on the first one.

File: src/gballoc.c [hunk 2]
```diff
@@ -117,7 +118,7 @@ void* gballoc_malloc(size_t size)
                 head = allocation;
 
                 g_allocations++;
-                totalSize += size;
+                totalSize = safe_add_size_t(totalSize, size);
```
Comment (ewertons) [inline src/gballoc.c:121]: This should be checked for size_t overflow.

File: src/gballoc.c [hunk 3]
```diff
@@ -168,12 +169,12 @@ void* gballoc_calloc(size_t nmemb, size_t size)
             {
                 /* Codes_SRS_GBALLOC_01_021: [If the underlying calloc call is successful, gballoc_calloc shall increment the total memory used with nmemb*size.] */
                 allocation->ptr = result;
-                allocation->size = nmemb * size;
+                allocation->size = safe_multiply_size_t(nmemb, size);
```
Comment (ewertons) [inline src/gballoc.c:172]: Should check for overflow.

File: src/gballoc.c [hunk 4]
```diff
@@ -168,12 +169,12 @@ void* gballoc_calloc(size_t nmemb, size_t size)
...
                 allocation->next = head;
                 head = allocation;
                 g_allocations++;
 
-                totalSize += allocation->size;
+                totalSize = safe_add_size_t(totalSize, allocation->size);
```
Comment (ewertons) [inline src/gballoc.c:177]: Should check for overflow.

File: src/gballoc.c [hunk 5]
```diff
@@ -266,7 +267,7 @@ void* gballoc_realloc(void* ptr, size_t size)
                 }
 
                 /* Codes_SRS_GBALLOC_01_007: [If realloc is successful, gballoc_realloc shall also increment the total memory used value tracked by this module.] */
-                totalSize += size;
+                totalSize = safe_add_size_t(totalSize, size);
```
Comment (ewertons) [inline src/gballoc.c:270]: Should check for overflow.

**Review by ewertons — APPROVED** [review]

File: adapters/tlsio_mbedtls.c [hunk 6]
```diff
@@ -342,9 +342,10 @@ static int on_io_recv(void *context, unsigned char *buf, size_t sz)
 
         if (result > 0)
         {
+            size_t read_byte_count = safe_subtract_size_t(tls_io_instance->socket_io_read_byte_count, (size_t)result);
             (void)memcpy((void *)buf, tls_io_instance->socket_io_read_bytes, result);
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply adapters/tlsio_mbedtls.c:346]: probably not going to get a on_io_recv with a zero byte count

File: adapters/tlsio_wolfssl.c [hunk 7]
```diff
@@ -407,9 +407,10 @@ static int on_io_recv(WOLFSSL *ssl, char *buf, int sz, void *context)
 
         if (result > 0)
         {
+            size_t read_byte_count = safe_subtract_size_t(tls_io_instance->socket_io_read_byte_count, (size_t)result);
             (void)memcpy(buf, tls_io_instance->socket_io_read_bytes, result);
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply adapters/tlsio_wolfssl.c:411]: probably not going to get a on_io_recv with a zero byte count

File: src/constbuffer_array_batcher.c [hunk 8]
```diff
@@ -46,11 +46,14 @@ CONSTBUFFER_ARRAY_HANDLE constbuffer_array_batcher_batch(CONSTBUFFER_ARRAY_HANDL
             /* Codes_SRS_CONSTBUFFER_ARRAY_BATCHER_01_003: [ Otherwise constbuffer_array_batcher_batch shall obtain the number of buffers used by each CONSTBUFFER_ARRAY. ]*/
 
             /* Codes_SRS_CONSTBUFFER_ARRAY_BATCHER_01_004: [ constbuffer_array_batcher_batch shall allocate memory for the header buffer (enough to hold the entire batch header namingly (count + 1) uint32_t values). ]*/
-            header_memory = malloc(sizeof(uint32_t) * ((size_t)count + 1));
-            if (header_memory == NULL)
+            size_t count_size_t = (size_t)count;  // handle -Werror=type-limits
+            size_t malloc_size = safe_add_size_t(count_size_t, 1);
```
[the remark this answers is not in the document]
Comment (ericwol-msft) [reply src/constbuffer_array_batcher.c:50]: no, got compiler error on GCC.

File: src/gballoc.c (hunk 3 above)
Comment (ericwol-msft) [reply src/gballoc.c:172]: not needed. This is only for malloc stats update

File: src/gballoc.c (hunk 4 above)
Comment (ericwol-msft) [reply src/gballoc.c:177]: not needed. This is only for malloc stats update

File: src/gballoc.c (hunk 5 above)
Comment (ericwol-msft) [reply src/gballoc.c:270]: not needed. This is only for malloc stats update

File: src/gballoc.c (hunk 2 above)
Comment (ericwol-msft) [reply src/gballoc.c:121]: not needed. This is only for malloc stats update

File: adapters/tlsio_mbedtls.c [hunk 9]
```diff
@@ -342,9 +342,11 @@ static int on_io_recv(void *context, unsigned char *buf, size_t sz)
...
-            (void)memmove(tls_io_instance->socket_io_read_bytes, tls_io_instance->socket_io_read_bytes + result, tls_io_instance->socket_io_read_byte_count - result);
-            tls_io_instance->socket_io_read_byte_count -= result;
+            size_t memmove_src = safe_add_size_t(tls_io_instance->socket_io_read_bytes, result);
+            (void)memmove(tls_io_instance->socket_io_read_bytes, (const void*)memmove_src, read_byte_count);
```
Comment (vaavva) [inline adapters/tlsio_mbedtls.c:348]: do read_byte_count and memove_src need to be checked for over/underflow?

File: src/azure_base64.c [hunk 10]
```diff
@@ -239,13 +240,13 @@ static STRING_HANDLE Base64_Encode_Internal(const unsigned char* source, size_t
     bool isBufferOverflow = false;
     char* encoded;
     size_t currentPosition = 0;
-    neededSize += (size == 0) ? (0) : ((((size - 1) / 3) + 1) * 4);
-    neededSize += 1; /*+1 because \0 at the end of the string*/
+    neededSize = safe_add_size_t(neededSize, (size == 0) ? (0) : ((((size - 1) / 3) + 1) * 4));
```
Comment (vaavva) [inline src/azure_base64.c:243]: Does the rest of the math in this line need safe math?

File: src/consolelogger.c [hunk 11]
```diff
@@ -15,7 +15,7 @@ static char* vprintf_alloc(const char* format, va_list va)
 {
     char* result;
     int neededSize = vsnprintf(NULL, 0, format, va);
-    if (neededSize < 0)
+    if (neededSize < 0 || (neededSize + 1) < 0)
```
Comment (vaavva) [inline src/consolelogger.c:18]: Is this a case of not wanting to include safe math in this file? (and in etwlogger_driver)

File: src/constbuffer_array_batcher.c [hunk 12]
```diff
@@ -46,11 +46,14 @@ CONSTBUFFER_ARRAY_HANDLE constbuffer_array_batcher_batch(CONSTBUFFER_ARRAY_HANDL
...
+            malloc_size = safe_multiply_size_t(malloc_size, sizeof(uint32_t));
+            if (malloc_size == SIZE_MAX ||
+                (header_memory = malloc(malloc_size)) == NULL)
             {
                 /* Codes_SRS_CONSTBUFFER_ARRAY_BATCHER_01_010: [ If any error occurrs, constbuffer_array_batcher_batch shall fail and return NULL. ]*/
-                LogError("malloc failed");
+                LogError("malloc failed, size:%zu", malloc_size);
```
Comment (vaavva) [inline src/constbuffer_array_batcher.c:56]: header_memory = NULL; needed?

File: src/strings.c [hunk 13]
```diff
@@ -444,7 +449,8 @@ int STRING_concat_with_STRING(STRING_HANDLE s1, STRING_HANDLE s2)
 
         size_t s1Length = strlen(dest->s);
         size_t s2Length = strlen(src->s);
-        char* temp = (char*)realloc(dest->s, s1Length + s2Length + 1);
+        size_t realloc_size = safe_add_size_t(safe_add_size_t(s1Length, s2Length), 1);
+        char* temp = (char*)realloc(dest->s, realloc_size);
         if (temp == NULL)
```
Comment (vaavva) [inline src/strings.c:454]:
```suggestion
        char* temp;
        if (realloc_size == SIZE_MAX ||
            (temp = (char*)realloc(dest->s, realloc_size)) == NULL)
```

**Review by vaavva — APPROVED** [review]

File: src/consolelogger.c (hunk 11 above)
Comment (ericwol-msft) [reply src/consolelogger.c:18]: yes

**Pull request merged**